pkg/validate/idempotence.go

```diff
@@ -0,0 +1,125 @@
+
+	"sigs.k8s.io/cri-tools/pkg/framework"
+)
+
+var _ = framework.KubeDescribe("Idempotence", func() {
+	f := framework.NewDefaultCRIFramework()
+	c := context.Background()
+
+	var (
+		rc internalapi.RuntimeService
+		ic internalapi.ImageManagerService
+	)
+
+	BeforeEach(func() {
+		rc = f.CRIClient.CRIRuntimeClient
+		ic = f.CRIClient.CRIImageClient
+	})
+
+	// https://github.com/kubernetes/cri-api/blob/c20fa40/pkg/apis/runtime/v1/api.proto#L45-L46
+	Context("StopPodSandbox", func() {
+		It("should not return an error if not found", func() {
+			By("stop not existing sandbox")
+			Expect(rc.StopPodSandbox(c, uuid.New().String())).NotTo(HaveOccurred())
+		})
+
+		It("should not return an error if already stopped", func() {
+			By("run PodSandbox")
+			podID := framework.RunDefaultPodSandbox(rc, "idempotence-pod-")
+
+			By("stop sandbox")
+			testStopPodSandbox(rc, podID)
+
+			By("stop sandbox again")
+			testStopPodSandbox(rc, podID)
+
+			By("delete sandbox as cleanup")
+			Expect(rc.RemovePodSandbox(c, podID)).NotTo(HaveOccurred())
+		})
+	})
+
+	// https://github.com/kubernetes/cri-api/blob/c20fa40/pkg/apis/runtime/v1/api.proto#L53-L54
+	Context("RemovePodSandbox", func() {
+		It("should not return an error if not found", func() {
+			By("remove not existing sandbox")
+			Expect(rc.RemovePodSandbox(c, uuid.New().String())).NotTo(HaveOccurred())
+		})
+	})
+
+	// https://github.com/kubernetes/cri-api/blob/c20fa40/pkg/apis/runtime/v1/api.proto#L67-L68
+	Context("StopContainer", func() {
+		It("should not return an error if not found", func() {
+			By("test stop not existing container")
+			Expect(rc.StopContainer(c, uuid.New().String(), 0)).NotTo(HaveOccurred())
+		})
+
+		It("should not return an error if already stopped", func() {
+			By("create sandbox")
+			podID, podConfig := framework.CreatePodSandboxForContainer(rc)
+
+			By("create container")
+			containerID := framework.CreatePauseContainer(rc, ic, podID, podConfig, "idempotence-container-")
+
+			By("start container")
+			startContainer(rc, containerID)
+
+			By("stop container")
+			testStopContainer(rc, containerID)
+
+			By("remove container")
+			removeContainer(rc, containerID)
+
+			By("remove container again")
+			removeContainer(rc, containerID)
+
+			By("stop sandbox as cleanup")
+			Expect(rc.StopPodSandbox(c, podID)).NotTo(HaveOccurred())
+
+			By("delete sandbox as cleanup")
+			Expect(rc.RemovePodSandbox(c, podID)).NotTo(HaveOccurred())
+		})
+	})
+
+	// https://github.com/kubernetes/cri-api/blob/c20fa40/pkg/apis/runtime/v1/api.proto#L74-L75
+	Context("RemoveContainer", func() {
+		It("should not return an error if not found", func() {
+			By("remove not existing comtainer")
+			Expect(rc.RemoveContainer(c, uuid.New().String())).NotTo(HaveOccurred())
+		})
+	})
+
+	// https://github.com/kubernetes/cri-api/blob/c20fa40/pkg/apis/runtime/v1/api.proto#L156-L157
+	Context("RemoveImage", func() {
+		It("should not return an error if not found", func() {
```

Comment:
why not to test deleting the same image twice?

Reply:
Sure, I can add this as well.